File: Medium/Java/1011. Capacity To Ship Packages Within D Days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int start = 0;
        int end = 0;

        for(int i : weights){
            start = Math.max(start, i);
            end += i;
        }

        while(start <= end){
            int mid = start + ((end - start) / 2);
            int cap = findDay(weights, mid);
            if(cap <= days)
                end = mid - 1;
            else
                start = mid + 1;
        }

        return start;
    }

    public int findDay(int[] weights, int mid){
        int date = 0;
        int sum = 0;

        for(int i = 0; i < weights.length; i++){
            if(sum + weights[i] <= mid){
                sum += weights[i];
            }else{
                sum = weights[i];
                date++;
            }
        }

        return date + 1;
    }
}

File: Medium/Python/1011. Capacity To Ship Packages Within D Days.py
class Solution(object):
    def shipWithinDays(self, weights, days):
        start = 0
        end = 0

        for i in weights:
            start = max(start, i)
            end += i
        
        while start <= end:
            mid = start + ((end - start) / 2)
            cap = self.findDay(weights, mid)
            if cap <= days:
                end = mid - 1
            else:
                start = mid + 1
        
        return start
    
    def findDay(self, weights, mid):
        date = 0
        tempSum = 0

        for i in range(len(weights)):
            if tempSum + weights[i] <= mid:
                tempSum += weights[i]
            else:
                tempSum = weights[i]
                date += 1
        
        return date + 1
